tests/framework/extensions/users/users.go

```diff
@@ -0,0 +1,95 @@
+package users
+
+import (
+	"strings"
+
+	"github.com/rancher/norman/types"
+	"github.com/rancher/rancher/pkg/api/scheme"
+	v3 "github.com/rancher/rancher/pkg/apis/management.cattle.io/v3"
+	"github.com/rancher/rancher/tests/framework/clients/rancher"
+	management "github.com/rancher/rancher/tests/framework/clients/rancher/generated/management/v3"
+	"github.com/rancher/rancher/tests/framework/pkg/wait"
+	"github.com/rancher/rancher/tests/integration/pkg/defaults"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/watch"
+)
+
+func CreateUserWithRole(rancherClient *rancher.Client, user *management.User, roles ...string) (*management.User, error) {
+	createdUser, err := rancherClient.Management.User.Create(user)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, role := range roles {
+		roleBinding := &management.GlobalRoleBinding{
+			GlobalRoleID: role,
+			UserID:       createdUser.ID,
+		}
+
+		_, err = rancherClient.Management.GlobalRoleBinding.Create(roleBinding)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return createdUser, nil
+}
+
+func AddProjectMember(rancherClient *rancher.Client, project *management.Project, user *management.User, projectRole string) error {
+	role := &management.ProjectRoleTemplateBinding{
+		ProjectID:       project.ID,
+		UserPrincipalID: user.PrincipalIDs[0],
```

Comment:
are users guaranteed to have at least 1 principal?

Reply:
Yes they are